### PR TITLE
Make Compose invocations reentrant

### LINKED ISSUE #419: Make Compose invocations reentrant by removing shared call state
## Summary

`Compose` stores invocation-specific bookkeeping on the configured pipeline and transform instances. Two overlapping calls to the same `Compose` object can overwrite that bookkeeping and let one sample finish with state belonging to another sample.

This issue defines the concurrency contract for `Compose` and removes the shared invocation state that violates it. PR #418 exposed the problem while adding a Tensor-to-NumPy compatibility route for mixed transform pipelines. The underlying limitation already predates that PR.

## Why this matters

The common `DataLoader(num_workers > 0)` configuration usually creates a separate dataset and `Compose` instance in each worker process. It does not share one Python object and therefore does not exercise this race.

The race appears when application code shares one configured pipeline across overlapping calls, for example through a `ThreadPoolExecutor`, a web service, a threaded data source, or `asyncio.to_thread`. CPython's GIL does not make a multi-step mutation/read/clear protocol atomic. NumPy and Torch operations can also run native work while the GIL is released.

The desired property is reentrancy: overlapping calls to the same `Compose` instance must produce results consistent with some serial order. Each result must keep its own representation, layout, annotations, applied configuration, and trace. A shared seed does not need to promise which caller receives the next random draw unless the caller imposes an ordering.

## Concrete race exposed by #418

PR #418 adds three fields that describe the current Tensor call:

- `_tensor_spatial_targets`
- `_tensor_annotation_targets`
- `_tensor_requires_numpy_bridge`

For a Tensor input whose pipeline contains a NumPy-only transform, `Compose._validate_tensor_inputs()` records the supplied targets and selects the one whole-pipeline NumPy bridge. `_bridge_tensor_data_to_numpy()` and the restoration helpers read those fields later in the same invocation.

The following interleaving is valid today:

```text
call A: Tensor image + Tensor mask enter a pipeline that needs the NumPy bridge
call A: records targets=(image, mask), requires_numpy_bridge=True

call B: NumPy image, or a Tensor call with only image, enters the same Compose object
call B: clears or replaces the three fields

call A: resumes and reads B's state
```

Call A can then skip its required conversion, convert only its image while leaving its mask as a Tensor, or skip conversion back to Tensor at the public output boundary. A NumPy-only transform can receive a Tensor, and an otherwise valid call can return inconsistent target types.

`try/finally` cleanup prevents state from leaking after a serial call raises. It cannot isolate two calls whose lifetimes overlap.

## Existing shared invocation state

The Tensor route is one instance of a broader pattern. The current code also stores these per-invocation values on `Compose`:

| State | Used for | Failure when another call replaces it |
| --- | --- | --- |
| `_added_channel_dim`, `_added_channel_canonical` | Restoring a grayscale target after preprocessing | A call can retain an added channel or remove one that belonged to another call. |
| `_instance_count`, `_repack_after_processors` | Rebuilding `instance_binding` outputs | A call can repack against another call's instance count or cleanup flags. |
| `_tensor_spatial_targets`, `_tensor_annotation_targets`, `_tensor_requires_numpy_bridge` | Choosing and undoing the Tensor compatibility bridge | A call can use another call's conversion plan and output type contract. |

`BasicTransform` has a related runtime-state problem. Each invocation writes `self.params` and `self.applied_config`; `Compose` reads `applied_config` after the transform returns when `save_applied_params=True`. Concurrent calls can therefore associate one sample's applied configuration with another sample. `run_with_trace()` has the same concern when it records transform state.

All configured transforms also share the pipeline's RNG objects. Sharing an RNG stream can be acceptable if calls are serialized and the contract permits the caller-dependent order of draws. It needs an explicit policy before claiming parallel execution or per-call reproducibility.

## Required scope and decision

The implementation must choose and document one of these contracts:

1. **Serialized concurrent calls.** Guard each outer `Compose.__call__()` and `Compose.run_with_trace()` invocation with a reentrant lock. This preserves the existing mutable pipeline model and produces a valid serial order. It does not provide parallel augmentation from one pipeline object.
2. **Fully isolated invocation contexts.** Pass a call-local context through `Compose`, nested compositions, processors, tracing, and transforms. The context owns representation planning, grayscale restoration, instance-binding bookkeeping, applied parameters, and RNG allocation. This enables genuine overlap, but it is a larger API and state-management redesign.

The first implementation may serialize calls. It must not present that as parallel augmentation. Callers that need parallel CPU work can use independent pipeline instances until an isolated-context design exists.

Regardless of the selected contract, PR #418 should keep its Tensor bridge plan in a call-local value, for example `TensorCallState`, rather than add another observable lifetime-dependent field to `Compose`. That targeted change prevents the PR from adding a new race while this issue tracks the whole object model.

## Implementation plan

1. Inventory every field written during `__call__()` or `run_with_trace()` on `Compose`, nested compositions, processors, and `BasicTransform`.
2. Classify each field as immutable configuration, reusable cache, random state, or invocation state.
3. Move the Tensor bridge plan in #418 to a local object passed explicitly to bridge and restore helpers.
4. Implement the selected concurrency contract across both normal and traced execution. The synchronization boundary must cover RNG synchronization, parameter sampling, transform application, bookkeeping, postprocessing, restoration, and cleanup.
5. Preserve serial behavior, serialization, replay, `save_applied_params`, additional targets, grayscale handling, and instance binding.
6. Document the public concurrency and seeded-RNG semantics in the `Compose` API and the Tensor CPU backend design document.

## Regression coverage

Add permanent tests that use a deterministic synchronization mechanism. A timing-based `sleep` test is insufficient.

- Run overlapping calls to the same `Compose` instance with Tensor fallback inputs whose target sets differ, such as `image + mask` and `image`.
- Assert that every caller receives its own expected Tensor/NumPy boundary type and spatial layout.
- Cover `__call__()` and `run_with_trace()`.
- Cover grayscale channel restoration and `instance_binding` bookkeeping, because they use existing shared invocation state.
- With `save_applied_params=True`, verify that each output records parameters sampled for that exact input.
- Define the seeded-RNG assertion from the selected contract. A serialized implementation should test a valid serial sequence; an isolated implementation should test the documented per-call stream rule.
- Run a focused threaded regression test plus the relevant Compose, Tensor, replay, tracing, and DataLoader suites. The DataLoader test verifies integration; it does not replace the shared-object threaded test.

## Completion criteria

- The public docs state whether one `Compose` object supports overlapping calls and define seeded behavior under that use.
- No invocation state can be overwritten by another overlapping invocation.
- Tensor fallback, grayscale restoration, instance binding, `save_applied_params`, and tracing satisfy the chosen contract.
- Regression tests fail against the vulnerable behavior and pass deterministically under the selected implementation.
- Existing serial Compose behavior and public input/output contracts remain unchanged.

## Out of scope

This issue does not add CUDA, MPS, autograd, or parallel kernel execution to the CPU Tensor backend. It also does not treat separate `DataLoader` worker processes as concurrent access to one Python `Compose` object.

### PR DESCRIPTION
**Pull request by aswanth-07**
## Summary

Fixes #419.

This implements the serialized concurrent-call contract described in the issue:

- guard each complete `Compose.__call__()` and `run_with_trace()` invocation with a process-local reentrant lock;
- keep Tensor bridge planning in a call-local `_TensorCallState` instead of mutable pipeline fields;
- apply the same invocation boundary to nested traced compositions and `ReplayCompose` finalization;
- exclude the runtime lock from pickle state and recreate it after unpickling; and
- document that overlapping calls are reentrant but serialized, with lock acquisition order determining the seeded RNG sequence.

The implementation does not claim parallel augmentation from one pipeline instance. Callers that need parallel CPU work should use independent `Compose` instances.

## Regression coverage

`tests/test_compose_reentrancy.py` uses `threading.Event` synchronization rather than sleeps to force a second caller to attempt entry while the first is inside transform dispatch. It covers:

- `Compose` and `ReplayCompose`;
- normal and traced execution;
- Tensor fallback with `image + mask` racing a NumPy `image` call;
- grayscale layout and Tensor/NumPy boundary restoration;
- `save_applied_params` and trace parameters;
- `instance_binding`; and
- seeded overlapping calls matching the lock-acquisition serial order.

The original implementation fails the threaded regression because both callers enter transform dispatch concurrently.

## Validation

- `python -m pytest tests/test_compose_reentrancy.py -q`: **7 passed**
- relevant Compose/Tensor/tracing/replay/worker/serialization suites: **1571 passed, 8 skipped**
- `python -m tools.quality_gate fast`: **passed**
  - Ruff format/check: passed
  - mypy: passed
  - Pyrefly: 0 errors (2 configured suppressions)
  - contract suite: **1276 passed**
  - CI matrix/shard, benchmark coverage, performance budget, legal integrity, regression vectors, pre-commit, and Compose smoke: passed
- explicit pre-commit run over all four changed/new files: passed
- `git diff --check`: passed

## Performance evidence

I ran the repository Tensor Compose benchmark before and after the change over the complete 3 sizes × 3 channel counts × 2 dtypes matrix (`repeats=7`, `iterations=128`, CPU, DataLoader skipped for the matrix). Every cell passed correctness.

- `numpy_with_to_tensor`: median cell delta **+1.67%**
- `tensor_direct`: median cell delta **+1.35%**

Because direct `NoOp` calls take only about 11–20 microseconds on this Windows host, some individual percentage deltas were noisy. I investigated every >5% cell with 15 repeats × 512 iterations. All but two reruns fell below 5%; the remaining deltas were +1.29 microseconds for NumPy-to-Tensor and +0.58 microseconds for direct Tensor, while equivalent cells moved in both directions.

A representative 256×256×3 uint8 DataLoader run (`workers=0`, 64 samples, batch size 8, 15 repeats) measured:

- NumPy + `ToTensorV2`: **+3.08%** per sample
- direct Tensor: **-12.49%** per sample

This change adds synchronization and call-local bookkeeping only. It does not add full-array passes, copies, conversions, kernels, LUTs, grouped reductions, or random draws. Vectorization, in-place operation, backend parity, and Albucore migration are therefore not applicable; the lock and invocation policy belong to `Compose`, not an image-processing primitive.

## AI assistance disclosure

OpenAI Codex assisted with the diagnosis, implementation, regression tests, documentation, validation, and benchmark analysis for this pull request. This assistance is explicitly disclosed for maintainer review.

## Summary by Sourcery

Make Compose and ReplayCompose invocations reentrant by serializing concurrent calls under a per-instance lock and making Tensor route planning call-local rather than mutable pipeline state.

New Features:
- Introduce a reentrant, process-local invocation lock on Compose and ReplayCompose to serialize overlapping __call__ and run_with_trace executions while preserving reentrancy.

Enhancements:
- Refactor Tensor/Numpy bridge planning into an immutable per-call _TensorCallState so Tensor route and annotation handling are local to each invocation.
- Align nested traced Compose and ReplayCompose execution with the same invocation boundary to keep tracing, replay finalization, and Tensor restoration consistent across concurrent calls.
- Ensure runtime locks are excluded from Compose pickle state and recreated on unpickling so worker processes do not share synchronization objects.
- Document the concurrent invocation and Tensor route ownership contracts for Compose, including the serialized RNG behavior under overlapping seeded calls.

Tests:
- Add threaded regression tests in tests/test_compose_reentrancy.py to validate serialized concurrent calls across normal execution, tracing, replay, Tensor/NumPy restoration, grayscale layout, applied parameters, instance binding, and seeded RNG ordering.

### CODE REVIEW

File: albumentations/core/composition.py [hunk 1]
```diff
@@ -1311,6 +1323,7 @@ def __init__(
             seed=seed,
             save_applied_params=save_applied_params,
         )
+        self._call_lock = threading.RLock()
 
         self.telemetry = telemetry
         self._resolve_processors(bbox_params, keypoint_params)
```
[the remark this answers is not in the document]
Comment (aswanth-07) [reply albumentations/core/composition.py:1329]: Updated `Compose.__call__` and `run_with_trace` to use `with self._call_lock:` while preserving the existing critical-section boundaries and nested `RLock` usage. The 1,571-test relevant suite and full fast quality gate pass.

File: tests/test_compose_reentrancy.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+class _BlockingAppliedConfigTransform(A.ImageOnlyTransform):
+    """Hold one invocation inside transform dispatch while detecting a second invocation."""
+
+    def __init__(self, marker: int = 0, p: float = 1.0):
+        super().__init__(p=p)
+        self.marker = marker
+        self.first_entered = Event()
+        self.second_entered = Event()
+        self.release_first = Event()
+
```
[the remark this answers is not in the document]
Comment (aswanth-07) [reply tests/test_compose_reentrancy.py:26]: Added `test_call_lock_and_tensor_state_recover_after_transform_exception`. It forces a Tensor call to fail inside the transform, then reuses the same pipeline for a successful NumPy call and verifies fresh applied parameters and representation state.

File: docs/design/compose-serialization-and-tracing.md [hunk 3]
```diff
@@ -78,6 +78,22 @@ dictionary. None of these runtime objects is serialized or retained in a DataLoa
...
 targets, or read a clock.
 
+## Concurrent invocation contract
+
+One configured `Compose` or `ReplayCompose` instance is reentrant through serialized execution. A process-local
+reentrant lock covers each complete `__call__()` and `run_with_trace()` invocation, including worker RNG
+synchronization, probability and parameter sampling, Tensor route selection, preprocessing, transform dispatch,
+processor work, tracing, restoration, replay finalization, and cleanup. Overlapping callers therefore observe results
+consistent with one serial order and cannot replace another call's grayscale, instance-binding, applied-parameter, or
+representation bookkeeping.
```
[the remark this answers is not in the document]
Comment (aswanth-07) [reply docs/design/compose-serialization-and-tracing.md]: Fixed the wording to “applied parameter”.

File: tests/test_compose_reentrancy.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    with ThreadPoolExecutor(max_workers=2) as executor:
+        first_future = executor.submit(_invoke, pipeline, first_mode, first_data)
+        second_future = None
+        try:
+            if not probe.first_entered.wait(timeout=5):
+                raise TimeoutError("First Compose invocation did not reach transform dispatch")
+            second_future = executor.submit(run_second)
+            if not second_attempted.wait(timeout=5):
+                raise TimeoutError("Second Compose invocation did not start")
+            second_overlapped = probe.second_entered.wait(timeout=0.25)
```
[the remark this answers is not in the document]
Comment (aswanth-07) [reply tests/test_compose_reentrancy.py]: Replaced the 250 ms transform-entry probe with an instrumented reentrant lock. The test now waits for an explicit event emitted when a different thread attempts to acquire the held invocation lock, eliminating the scheduler-sensitive false-negative window.

**Comment by ternaus** [comment]
It is a bit more complex than that, implemented it in other PR.

**Comment by aswanth-07** [comment]
Alright, I’ll see if there’s anything else I can contribute to :)
